tensorizer/_linear_partition.py

```diff
@@ -51,108 +45,177 @@ def partition(
+    for w in weights:
+        if w < 0:
+            raise ValueError("All weights must be non-negative")
+    prefix_sums: Tuple[int, ...] = tuple(
+        itertools.accumulate(weights, initial=0)
     )
 
-    # Each row represents a different prefix
-    # No new partitions means no splitting; the total weight is the whole prefix
-    best_prefix_partition[:, 0] = prefix_sums
-
-    # Each column represents splitting into a different number of partitions
-    # Since the first row is a prefix with only one item,
-    # regardless of the number of partitions, it will have the same maximum
-    # weight (that single weight)
-    best_prefix_partition[0, :] = weights[0]
-
-    partition_starts: np.ndarray = np.empty(
-        (n - 1, partitions - 1), dtype=np.uint64
-    )
-    # The columns in partition_starts mean the same as the columns in
-    # best_prefix_partition, except row & column 0 would be meaningless,
-    # so we remove them, since there's nothing to end if there are no new
-    # partitions allowed, and nothing precedes position 0.
-    # The rows in partition_starts are a lookup table
-    # An entry at partition_starts[end, k] being `start` means the optimal
-    # partition ending immediately before `end` starts at `start`,
-    # with k other partitions behind it.
-    # The final partition indices come from iterated application of
-    # looking up `end` in partition_starts, and replacing it with the result
-    for end in range(1, n):
-        for partitions_remaining in range(1, partitions):
-            for start in range(end):
-                # If you form a new partition between `start` and `end`,
-                # then the largest sum overall is either the new one,
-                # or the previous largest sum in everything preceding `start`
-                # (with fewer allowed partitions)
-                largest_partition = max(
-                    best_prefix_partition[start, partitions_remaining - 1],
-                    prefix_sums[end] - prefix_sums[start],
-                )
-                best = best_prefix_partition[end, partitions_remaining]
-                if (
-                    prefer_fewer
-                    and largest_partition < best
-                    or not prefer_fewer
-                    and largest_partition <= best
-                ):
-                    # If forming a partition here gave
-                    # a record low outcome, record it.
-                    # If prefer_fewer is true, this may result in fewer
-                    # partitions than originally requested, if adding more
-                    # wouldn't reduce the maximum partition size either way
-                    best_prefix_partition[end, partitions_remaining] = (
-                        largest_partition
+    inf: int = prefix_sums[-1] + 1
+    sentinel = (inf, inf)
+    memo = [sentinel] * ((n + 1) * partitions)
+    # Key function: end * partitions + preceding_parts
+    null = (0, 0)
```

Comment:
I know that this isn't a keyword in Python, but seeing `null` defined gives me hives. :joy: